### ISSUE TITLE
[Shift-JIS problem ?] When Download folder as ZIP file, the Extracted ZIP file contains Garbled characters files and dirs name

### ISSUE THREAD

**Issue by falconws** [owncloud/core#24900]
### Steps to reproduce
1. Upload Folder named Japanese Character(for example "報告書/" dir)
2. Upload Files named Japanese Character(for example "テスト1.txt", "テスト2.txt") 
3. Download "報告書/" dir As ZIP file("報告書.zip")
4. Extract "報告書.zip" Windows Explorer

### Expected behaviour
Appear Garbled characters files and dirs name.

### Actual behaviour
Actual dirs and files name.

### Server configuration
**Operating system**: Ubuntu 14.04 LTS

**Web server:** Apache 2

**Database:** PostgreSQL 9

**PHP version:** 5

**ownCloud version:** (see ownCloud admin page) 8.2.3 (stable)

**Updated from an older ownCloud or fresh install:** fresh install

**Where did you install ownCloud from:**
https://download.owncloud.org/download/repositories/8.2/Ubuntu_14.04/

**Signing status (ownCloud 9.0 and above):** 
Unfortunately, Redirect to apps/files/ .

**List of activated apps:**

```
Enabled:
  - activity: 2.1.4
  - files: 1.2.1
  - files_pdfviewer: 0.7
  - files_sharing: 0.7.0
  - files_texteditor: 2.0
  - files_trashbin: 0.7.0
  - files_versions: 1.1.0
  - files_videoviewer: 0.1.3
  - firstrunwizard: 1.1
  - gallery: 14.2.0
  - notifications: 0.1.0
  - provisioning_api: 0.3.0
  - templateeditor: 0.1
Disabled:
  - encryption
  - external
  - files_external
  - user_external
  - user_ldap
```

**The content of config/config.php:**

```
{
    "system": {
        "instanceid": "ocjxqnxiywka",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "***FQDN***"
        ],
        "datadirectory": "\/var\/www\/owncloud\/data",
        "overwrite.cli.url": "https:\/\/***FQDN***\/owncloud",
        "dbtype": "pgsql",
        "version": "8.2.3.2",
        "dbname": "owncloud",
        "dbhost": "localhost",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "UTC",
        "installed": true
    }
}
```

**Are you using external storage, if yes which one:** local disk only

**Are you using encryption:** no

**Are you using an external user-backend.** no

### Client configuration
**Browser:** Google Chrome

**Operating system:** Windows 7

#### Web server error log
No log available.

#### ownCloud log (data/owncloud.log)
I don't want to paste owncloud.log because this log include sensitive data (for example, IP addr etc..)
and, my owncloud.log file is 24 Mega Byte, so can't paste all log.
![screenshot at 2016-05-30 23 29 24](https://cloud.githubusercontent.com/assets/2891031/15652060/8c7503a6-26be-11e6-9cc4-c9f5ec3346aa.png)

**Comment by falconws** [comment]
Sorry mistake.

**Comment by PVince81** [comment]
Could be either a bug in the ZIP streamer library https://github.com/McNetic/PHPZipStreamer or the client you are using to unzip isn't compatible. (CC @McNetic)

From what I remember Windows Explorer used to have problems with Zip files, but that was with XP. Not sure about Windows 7's zip folder.

Are you able to extract the ZIP files with other tools ?

**Comment by falconws** [comment]
We were test here.

---

Lhaplus: OK

---

Windows 7 Explorer Zip: NG
+Lhaca: NG
Linux(Ubuntu 14.04) unzip command: NG

---

I think this is bug.
zipped file name character code problem ?

Fix Concept below.
```php
$zip->addFile(mb_convert_encoding($zipStream, "SJIS-win", "UTF-8"));
```

**Comment by PVince81** [comment]
We cannot use the fix you proposed because it would break compatibility with other systems where people use UTF-8 or other encodings. It is by design that ownCloud uses UTF-8 everywhere.
Not sure why Windows cannot cope with UTF-8 encodings.

However, the `unzip` command from Ubuntu should work.

Can you double check that your ownCloud Linux environment is running with the locale "en_US.UTF-8" ? This is required to make sure encoding works properly everywhere. Also make sure you have the PHP module "intl" installed.

**Comment by falconws** [comment]
@PVince81 
Thank you for advice.

> Can you double check that your ownCloud Linux environment is running with the locale "en_US.UTF-8" ? 

We are Japanese so don't set "en_US".
But, use "UTF-8". See below.
```
# echo $LANG
ja_JP.UTF-8
```
> Also make sure you have the PHP module "intl" installed.
Maybe already installed.
```
# dpkg -l | grep php5-intl
ii  php5-intl                           5.5.9+dfsg-1ubuntu4.16           amd64        internationalisation module for php5
```

By the way, I found this.
https://github.com/ukitiyan/zipstream_ja/

This problem come from multibyte character...

**Comment by PVince81** [comment]
@falconws can you try changing the server locale to en_US.UTF-8 ?
You probably don't need to change it system-wide but just make sure the Apache/PHP process runs with that locale. See whether this fixes your issue.

**Comment by falconws** [comment]
@PVince81 I did your advice.

```
# service apache2 stop
 * Stopping web server apache2
# LANG=en_US.UTF-8 service apache2 start
 * Starting web server apache2
```

Unfortunately this issue still appear.

**Comment by PVince81** [comment]
@falconws okay, thanks for your effort.

I'll have a try locally, this looks suspicious.
Last time I tried with Ubuntu's `unzip` it used to work, that was with OC 8.1 or so.

**Comment by falconws** [comment]
For your reference, put broken zip with this issue.
And my unzip command details.

```
$ LANG=C dpkg -l unzip
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                        Version            Architecture       Description
+++-===========================-==================-==================-===========================================================
ii  unzip                       6.0-10~5.52~ja2    amd64              De-archiver for .zip files
```

with this unzip command extract attached file log below.
```
$ unzip テスト.zip
Archive:  テスト.zip
warning [テスト.zip]:  zipfile claims to be last disk of a multi-part archive;
  attempting to process anyway, assuming all parts have been concatenated
  together in order.  Expect "errors" and warnings...true multi-part support
  doesn't exist yet (coming soon).
error [テスト.zip]:  missing 8589934108 bytes in zipfile
  (attempting to process anyway)
error [テスト.zip]:  attempt to seek before beginning of zipfile
  (please check that you have transferred or created the zipfile in the
  appropriate BINARY mode and that you have compiled UnZip properly)
```
I hope this problem fix.

Thanks a lot.
[テスト.zip](https://github.com/owncloud/core/files/291069/default.zip)

**Comment by PVince81** [comment]
Oh, it looks like your issue is even worse than a simple encoding issue. Instead of extracting files where only the names are corrupt, it looks like your whole zip file is corrupt!

**Comment by falconws** [comment]
!?
make this zip file below.

* logged in ownCloud
* Click Left panel [All files]
* New -> Folder -> "テスト"
* Click "テスト" Folder
* New -> File -> "新規のテキストファイル作成.txt"
* Back to Parent Folder
* Download "テスト" Folder As "テスト.zip"

Why this zip file corrupted ?

**Comment by PVince81** [comment]
Just tried this and it worked fine with v8.2.3 (your version):
![japanese-zip](https://cloud.githubusercontent.com/assets/277525/15675552/c6dedb8c-2742-11e6-8ba3-2115c21f123b.png)

Also works fine on the CLI:
```
  % unzip テスト.zip
Archive:  テスト.zip
   creating: テスト/
 extracting: テスト/新規のテキストファイル作成.txt
```

Just in case, here are my versions:
- openSUSE Tumbleweed
- apache2-2.4.20-3.1.x86_64
- php5-5.6.21-1.1.x86_64
- unzip-6.00-29.6.x86_64

**Comment by PVince81** [comment]
@falconws here is the zip file generated from my instance: https://s3.owncloud.com/owncloud/index.php/s/MmEwI4fuQ7C4p4a

Can you check whether your system is able to extract it ?

This will help find out whether the problem is in your clients (unzip, etc) or whether your server is somehow not properly creating the zip file.

**Comment by falconws** [comment]
I was downloaded your "テスト.zip".
result:
Ubuntu 14.04 LTS's unzip command: NG (the same output)
Windows 7's Zip Explorer(standard feature): NG(the same 
https://cloud.githubusercontent.com/assets/2891031/15652060/8c7503a6-26be-11e6-9cc4-c9f5ec3346aa.png )

ownCloud doesn't support Japanese Client ?
Especially, use case Windows 7's standard feature(Zip Explorer) .

**Comment by PVince81** [comment]
Your Ubuntu has the same `unzip` version, I'd expect it to work there.

ownCloud always uses UTF-8 encodings for everything as it is expected to work on most systems.
Not sure why it doesn't work on both your Ubuntu and Windows.
It should **at least** work on Ubuntu.

**Comment by PVince81** [comment]
Does it work with p7zip ?
See http://askubuntu.com/a/744462

**Comment by falconws** [comment]
This issue appear download folder as zip only.
Simply, single file download(include Japanese character) is all OK(Windows and Ubuntu).

In Japan, for support Shift-JIS filename in zip We setup This PPA
https://launchpad.net/~japaneseteam/+archive/ubuntu/ppa

Details (Sorry Japanese only page. But you can understand Overview by seeing pictures) :
https://wiki.ubuntulinux.jp/Develop/UnzipTestcase

In other words, my unzip command is modified by Ubuntu Japanese Team for support Shift-JIS filename in zip.

**Comment by falconws** [comment]
Why only when download folder as zip garbled characters ?
Why no problem when download single file ?

Probably funny ownCloud use all of the scene use UTF-8.

**Comment by PVince81** [comment]
I'm sorry, this is nothing that can be fixed in ownCloud. ownCloud can only work properly with UTF-8.
Using other locale would cause a lot of other issues and also remove compatibility with other systems.

Unfortunately it seems that to work with non-UTF-8 clients the ZIP files would require using a different encoding. But it would also mean that such ZIP files cannot be opened on other UTF-8 systems.

Since I see that even your Ubuntu needed patching for proper support, you'll have to patch your ownCloud instance like you suggested here https://github.com/owncloud/core/issues/24900#issuecomment-222524851

We can't put such a change permanently in ownCloud.

CC @DeepDiver1975

**Comment by falconws** [comment]
OK...
I understand you said.

Thank you relationship my poor English.

**Comment by falconws** [comment]
oh!!
I'm happy because I applyed this Microsoft Patch
https://support.microsoft.com/en-us/kb/2704299

Windows Explorer zip feature make correct filename !!

**Comment by PVince81** [comment]
@falconws :smile:
